pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/MissingEqualsOnComparableRule.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule.errorprone;
+
+import net.sourceforge.pmd.lang.java.ast.ASTAnonymousClassDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTClassDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTRecordDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTTypeDeclaration;
+import net.sourceforge.pmd.lang.java.ast.internal.JavaAstUtils;
+import net.sourceforge.pmd.lang.java.rule.AbstractJavaRulechainRule;
+import net.sourceforge.pmd.lang.java.types.TypeTestUtil;
+
+public class MissingEqualsOnComparableRule extends AbstractJavaRulechainRule {
+
+    public MissingEqualsOnComparableRule() {
+        super(ASTClassDeclaration.class,
+                ASTRecordDeclaration.class,
+                ASTAnonymousClassDeclaration.class);
+    }
+
+    @Override
+    public Object visit(ASTAnonymousClassDeclaration node, Object data) {
+        visitTypeDecl(node, data);
+        return null;
+    }
+
+    @Override
+    public Object visit(ASTClassDeclaration node, Object data) {
+        visitTypeDecl(node, data);
+        return null;
+    }
+
+    @Override
+    public Object visit(ASTRecordDeclaration node, Object data) {
+        visitTypeDecl(node, data);
```

Comment:
Should this rule apply to records? One reason to use a record is to get `equals` and `hashCode` for free and it's quite likely that custom implementation of `compare` is compatible with those methods.

Reply:
yes, thx. covered as well.

Reply:
```suggestion
        return null;
```
then could remove it from the inspection list, right? 

If we dont care for it.

Reply:
> where some boilerplates are generated

Assuming this case only makes sense, if the record impl. differs in eq and hc from the default class impl.
Im not sure about this.

Reply:
TBC in dedicated issue.

Reply:
A class can have either the default `equals` (identity-based) or an overriden one (property-based).
Having a property-based `compareTo` and the identity-based `equals` is most likely a bug and is reported correctly by this new rule. 
But record classes have property-based `equals`, so the only thing to verify is that `equals` and `compareTo` take the same properties of the object into account. That would be nice, but it's not something that can be done reliably and it's also not enforced for non-record classes.

Reply:
> lombok or other annotation bases solution would ensure unity, but thats very specific. In Kotlin [Data classes](https://kotlinlang.org/docs/data-classes.html) might solve this issue.

No, they don't. They only care about unity of equals and hashCode. But as soon as you add a custom compareTo method, you need to be careful.
If you don't implement Comparable, then you are safe (regardless whether you use lombok, records, data classes).